File: scripts/syndrql-training/compiler/compiler.go
package compiler

import (
	"fmt"
	"strings"

	"syndrql-training/ir"
)

// Compile compiles an IR Statement to a SyndrQL string.
func Compile(stmt *ir.Statement) (string, error) {
	switch stmt.StatementType {
	// DML
	case "select":
		return compileSelectStmt(stmt.Select)
	case "insert":
		return compileInsert(stmt.Insert)
	case "update":
		return compileUpdate(stmt.Update)
	case "delete":
		return compileDelete(stmt.Delete)

	// DDL - Bundle
	case "createBundle":
		return compileCreateBundle(stmt.CreateBundle)
	case "dropBundle":
		return compileDropBundle(stmt.DropBundle)
	case "updateBundle":
		return compileUpdateBundle(stmt.UpdateBundle)

	// DDL - Index
	case "createIndex":
		return compileCreateIndex(stmt.CreateIndex)

	// DDL - View
	case "createView":
		return compileCreateView(stmt.CreateView)
	case "dropView":
		return compileDropView(stmt.DropView)
	case "refreshView":
		return compileRefreshView(stmt.RefreshView)

	// Transaction
	case "beginTransaction":
		return "BEGIN TRANSACTION;", nil
	case "commit":
		return "COMMIT;", nil
	case "rollback":
		return "ROLLBACK;", nil
	case "savepoint":
		return compileSavepoint(stmt.Savepoint)
	case "rollbackToSavepoint":
		return compileRollbackToSavepoint(stmt.RollbackToSavepoint)

	// Cursor
	case "declareCursor":
		return compileDeclareCursor(stmt.DeclareCursor)
	case "fetchCursor":
		return compileFetchCursor(stmt.FetchCursor)
	case "closeCursor":
		return compileCloseCursor(stmt.CloseCursor)

	// Prepared
	case "prepare":
		return compilePrepare(stmt.Prepare)
	case "execute":
		return compileExecute(stmt.Execute)
	case "deallocate":
		return compileDeallocate(stmt.Deallocate)

	// RBAC
	case "createUser":
		return compileCreateUser(stmt.CreateUser)
	case "updateUser":
		return compileUpdateUser(stmt.UpdateUser)
	case "deleteUser":
		return compileDeleteUser(stmt.DeleteUser)
	case "createRole":
		return compileCreateRole(stmt.CreateRole)
	case "updateRole":
		return compileUpdateRole(stmt.UpdateRole)
	case "deleteRole":
		return compileDeleteRole(stmt.DeleteRole)
	case "grantPermission":
		return compileGrantPermission(stmt.GrantPermission)
	case "grantRole":
		return compileGrantRole(stmt.GrantRole)
	case "revokePermission":
		return compileRevokePermission(stmt.RevokePermission)
	case "revokeRole":
		return compileRevokeRole(stmt.RevokeRole)

	// Database
	case "useDatabase":
		return compileUseDatabase(stmt.UseDatabase)
	case "createDatabase":
		return compileCreateDatabase(stmt.CreateDatabase)
	case "dropDatabase":
		return compileDropDatabase(stmt.DropDatabase)
	case "renameDatabase":
		return compileRenameDatabase(stmt.RenameDatabase)
	case "attachDatabase":
		return compileAttachDatabase(stmt.AttachDatabase)

	// Utility
	case "showDatabases":
		return "SHOW DATABASES;", nil
	case "showBundles":
		return "SHOW BUNDLES;", nil
	case "showBundle":
		return compileShowBundle(stmt.ShowBundle)
	case "showUsers":
		return "SHOW USERS;", nil
	case "showSessions":
		return "SHOW SESSIONS;", nil
	case "showViews":
		return compileShowViews(stmt.ShowViews)
	case "describeView":
		return compileDescribeView(stmt.DescribeView)
	case "showMigrations":
		return compileShowMigrations(stmt.ShowMigrations)
	case "showRateLimit":
		return "SHOW RATE LIMIT;", nil
	case "showVersions":
		return compileShowVersions(stmt.ShowVersions)
	case "showActiveSnapshots":
		return "SHOW ACTIVE SNAPSHOTS;", nil
	case "showConflictLog":
		return "SHOW CONFLICT LOG;", nil
	case "explainSelect":
		return compileExplainSelect(stmt.ExplainSelect)

	// Migration
	case "startMigration":
		return compileStartMigration(stmt.StartMigration)
	case "applyMigration":
		return compileApplyMigration(stmt.ApplyMigration)
	case "validateMigration":
		return compileValidateMigration(stmt.ValidateMigration)
	case "validateRollback":
		return compileValidateRollback(stmt.ValidateRollback)

	// Ops
	case "checkpoint":
		return "CHECKPOINT;", nil
	case "backupDatabase":
		return compileBackupDatabase(stmt.BackupDatabase)
	case "restoreDatabase":
		return compileRestoreDatabase(stmt.RestoreDatabase)
	case "invalidateSession":
		return compileInvalidateSession(stmt.InvalidateSession)

	default:
		return "", fmt.Errorf("unknown statement type: %s", stmt.StatementType)
	}
}

// CompileAll compiles all statements in an IREnvelope.
func CompileAll(env *ir.IREnvelope) ([]string, error) {
	var results []string
	for i, stmt := range env.Statements {
		sql, err := Compile(&stmt)
		if err != nil {
			return nil, fmt.Errorf("statement %d: %w", i, err)
		}
		results = append(results, sql)
	}
	return results, nil
}

// CompileSelect compiles a SelectStmt to SyndrQL. Exported for reuse by
// view/cursor/prepare compilers.
func CompileSelect(s *ir.SelectStmt) string {
	if s == nil {
		return ""
	}
	return compileSelectInner(s) + ";"
}

func compileSelectStmt(s *ir.SelectStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil select statement")
	}
	return CompileSelect(s), nil
}

func compileSelectInner(s *ir.SelectStmt) string {
	var b strings.Builder

	b.WriteString("SELECT ")
	if s.Distinct {
		b.WriteString("DISTINCT ")
	}

	// Fields
	if len(s.Fields) == 0 {
		b.WriteString("*")
	} else {
		for i, f := range s.Fields {
			if i > 0 {
				b.WriteString(", ")
			}
			b.WriteString(CompileSelectFieldExpr(f.Expression))
			if f.Alias != "" {
				b.WriteString(" AS ")
				b.WriteString(f.Alias)
			}
		}
	}

	// FROM
	b.WriteString(fmt.Sprintf(` FROM "%s"`, s.Bundle))

	// JOIN
	for _, j := range s.Joins {
		b.WriteString(" ")
		b.WriteString(j.JoinType)
		b.WriteString(fmt.Sprintf(` JOIN "%s" ON `, j.Bundle))
		for ci, c := range j.Conditions {
			if ci > 0 {
				b.WriteString(" AND ")
			}
			b.WriteString(fmt.Sprintf(`"%s"."%s" %s "%s"."%s"`,
				c.LeftBundle, c.LeftField, c.Operator, c.RightBundle, c.RightField))
		}
	}

	// WHERE
	if s.Where != nil {
		b.WriteString(" WHERE ")
		b.WriteString(CompileExpr(s.Where))
	}

	// GROUP BY
	if len(s.GroupBy) > 0 {
		b.WriteString(" GROUP BY ")
		b.WriteString(strings.Join(s.GroupBy, ", "))
	}

	// HAVING
	if s.Having != nil {
		b.WriteString(" HAVING ")
		b.WriteString(CompileExpr(s.Having))
	}

	// ORDER BY
	if len(s.OrderBy) > 0 {
		b.WriteString(" ORDER BY ")
		for i, o := range s.OrderBy {
			if i > 0 {
				b.WriteString(", ")
			}
			b.WriteString(o.Field)
			b.WriteString(" ")
			b.WriteString(o.Direction)
		}
	}

	// LIMIT
	if s.Limit > 0 {
		b.WriteString(fmt.Sprintf(" LIMIT %d", s.Limit))
	}

	// OFFSET
	if s.Offset > 0 {
		b.WriteString(fmt.Sprintf(" OFFSET %d", s.Offset))
	}

	// FOR UPDATE
	if s.ForUpdate {
		b.WriteString(" FOR UPDATE")
	}

	return b.String()
}

// compileValue compiles a Value to its SyndrQL representation.
func compileValue(v *ir.Value) string {
	return compileLiteral(v)
}

func compileInsert(s *ir.InsertStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil insert statement")
	}

	var b strings.Builder
	b.WriteString(fmt.Sprintf(`ADD DOCUMENT TO BUNDLE "%s" WITH (`, s.Bundle))

	for i, f := range s.Fields {
		if i > 0 {
			b.WriteString(", ")
		}
		b.WriteString(fmt.Sprintf("{%s = %s}", f.Name, compileValue(f.Value)))
	}

	b.WriteString(");")
	return b.String(), nil
}

func compileUpdate(s *ir.UpdateStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil update statement")
	}

	var b strings.Builder
	b.WriteString(fmt.Sprintf(`UPDATE DOCUMENTS IN BUNDLE "%s" (`, s.Bundle))

	for i, f := range s.Fields {
		if i > 0 {
			b.WriteString(", ")
		}
		b.WriteString(fmt.Sprintf("%s = %s", f.Name, compileValue(f.Value)))
	}

	b.WriteString(")")

	if s.Confirmed {
		b.WriteString(" CONFIRMED")
	}

	if s.Where != nil {
		b.WriteString(" WHERE ")
		b.WriteString(CompileExpr(s.Where))
	}

	b.WriteString(";")
	return b.String(), nil
}

func compileDelete(s *ir.DeleteStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil delete statement")
	}

	var b strings.Builder
	b.WriteString(fmt.Sprintf(`DELETE DOCUMENTS FROM "%s"`, s.Bundle))

	if s.Confirmed {
		b.WriteString(" CONFIRMED")
	}

	if s.Where != nil {
		b.WriteString(" WHERE ")
		b.WriteString(CompileExpr(s.Where))
	}

	b.WriteString(";")
	return b.String(), nil
}

func compileCreateBundle(s *ir.CreateBundleStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil create bundle statement")
	}

	var b strings.Builder
	b.WriteString(fmt.Sprintf(`CREATE BUNDLE "%s" WITH FIELDS (`, s.Bundle))

	for i, f := range s.Fields {
		if i > 0 {
			b.WriteString(", ")
		}
		b.WriteString(fmt.Sprintf(`{"%s", "%s", %t, %t`, f.Name, f.Type, f.Required, f.Unique))
		if f.DefaultValue != nil {
			b.WriteString(", ")
			b.WriteString(compileValue(f.DefaultValue))
		}
		b.WriteString("}")
	}

	b.WriteString(");")
	return b.String(), nil
}

func compileDropBundle(s *ir.DropBundleStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil drop bundle statement")
	}
	result := fmt.Sprintf(`DROP BUNDLE "%s"`, s.Bundle)
	if s.Force {
		result += " FORCE"
	}
	return result + ";", nil
}

func compileUpdateBundle(s *ir.UpdateBundleStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil update bundle statement")
	}

	var b strings.Builder
	b.WriteString(fmt.Sprintf(`UPDATE BUNDLE "%s"`, s.Bundle))

	if s.NewBundleName != "" {
		b.WriteString(fmt.Sprintf(` RENAME TO "%s"`, s.NewBundleName))
	}

	for _, c := range s.Changes {
		switch c.Action {
		case "ADD":
			if c.FieldDef != nil {
				b.WriteString(fmt.Sprintf(` ADD FIELD "%s" {"%s", "%s", %t, %t}`,
					c.Field, c.FieldDef.Name, c.FieldDef.Type, c.FieldDef.Required, c.FieldDef.Unique))
			}
		case "REMOVE":
			b.WriteString(fmt.Sprintf(` REMOVE FIELD "%s"`, c.Field))
		case "MODIFY":
			if c.FieldDef != nil {
				b.WriteString(fmt.Sprintf(` MODIFY FIELD "%s" {"%s", "%s", %t, %t}`,
					c.Field, c.FieldDef.Name, c.FieldDef.Type, c.FieldDef.Required, c.FieldDef.Unique))
			}
		}
	}

	b.WriteString(";")
	return b.String(), nil
}

func compileCreateIndex(s *ir.CreateIndexStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil create index statement")
	}

	var b strings.Builder
	b.WriteString(fmt.Sprintf(`CREATE %s "%s" ON BUNDLE "%s" WITH FIELDS (`,
		s.IndexType, s.IndexName, s.Bundle))

	for i, f := range s.Fields {
		if i > 0 {
			b.WriteString(", ")
		}
		b.WriteString(fmt.Sprintf(`{"%s", %t, %t}`, f.Name, f.Required, f.Unique))
	}

	b.WriteString(");")
	return b.String(), nil
}

func compileCreateView(s *ir.CreateViewStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil create view statement")
	}

	var b strings.Builder
	if s.IsMaterialized {
		b.WriteString("CREATE MATERIALIZED VIEW ")
	} else {
		b.WriteString("CREATE VIEW ")
	}
	b.WriteString(fmt.Sprintf(`"%s" AS `, s.ViewName))
	b.WriteString(CompileSelect(s.Query))
	return b.String(), nil
}

func compileDropView(s *ir.DropViewStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil drop view statement")
	}
	if s.IsMaterialized {
		return fmt.Sprintf(`DROP MATERIALIZED VIEW "%s";`, s.ViewName), nil
	}
	return fmt.Sprintf(`DROP VIEW "%s";`, s.ViewName), nil
}

func compileRefreshView(s *ir.RefreshViewStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil refresh view statement")
	}
	return fmt.Sprintf(`REFRESH MATERIALIZED VIEW "%s";`, s.ViewName), nil
}

func compileSavepoint(s *ir.SavepointStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil savepoint statement")
	}
	return fmt.Sprintf(`SAVEPOINT "%s";`, s.Name), nil
}

func compileRollbackToSavepoint(s *ir.RollbackToSavepointStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil rollback to savepoint statement")
	}
	return fmt.Sprintf(`ROLLBACK TO SAVEPOINT "%s";`, s.Name), nil
}

func compileDeclareCursor(s *ir.DeclareCursorStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil declare cursor statement")
	}
	inner := compileSelectInner(s.Query)
	return fmt.Sprintf("DECLARE %s CURSOR FOR %s;", s.CursorName, inner), nil
}

func compileFetchCursor(s *ir.FetchCursorStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil fetch cursor statement")
	}
	if s.FetchAll {
		return fmt.Sprintf("FETCH ALL FROM %s;", s.CursorName), nil
	}
	if s.Count == 1 {
		return fmt.Sprintf("FETCH NEXT FROM %s;", s.CursorName), nil
	}
	return fmt.Sprintf("FETCH %d FROM %s;", s.Count, s.CursorName), nil
}

func compileCloseCursor(s *ir.CloseCursorStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil close cursor statement")
	}
	return fmt.Sprintf("CLOSE %s;", s.CursorName), nil
}

func compilePrepare(s *ir.PrepareStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil prepare statement")
	}
	inner := compileSelectInner(s.Query)
	return fmt.Sprintf("PREPARE %s AS %s;", s.StatementName, inner), nil
}

func compileExecute(s *ir.ExecuteStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil execute statement")
	}
	return fmt.Sprintf("EXECUTE %s;", s.StatementName), nil
}

func compileDeallocate(s *ir.DeallocateStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil deallocate statement")
	}
	return fmt.Sprintf("DEALLOCATE %s;", s.StatementName), nil
}

func compileCreateUser(s *ir.CreateUserStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil create user statement")
	}
	return fmt.Sprintf(`CREATE USER "%s" WITH PASSWORD "%s";`, s.Username, s.Password), nil
}

func compileUpdateUser(s *ir.UpdateUserStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil update user statement")
	}
	result := fmt.Sprintf(`UPDATE USER "%s" SET PASSWORD = "%s"`, s.Username, s.NewPassword)
	if s.Force {
		result += " FORCE"
	}
	return result + ";", nil
}

func compileDeleteUser(s *ir.DeleteUserStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil delete user statement")
	}
	result := fmt.Sprintf(`DELETE USER "%s"`, s.Username)
	if s.Force {
		result += " FORCE"
	}
	return result + ";", nil
}

func compileCreateRole(s *ir.CreateRoleStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil create role statement")
	}
	result := fmt.Sprintf(`CREATE ROLE "%s"`, s.RoleName)
	if s.Description != "" {
		result += fmt.Sprintf(` WITH DESCRIPTION "%s"`, s.Description)
	}
	return result + ";", nil
}

func compileUpdateRole(s *ir.UpdateRoleStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil update role statement")
	}
	result := fmt.Sprintf(`UPDATE ROLE "%s" SET DESCRIPTION = "%s"`, s.RoleName, s.NewDescription)
	if s.Force {
		result += " FORCE"
	}
	return result + ";", nil
}

func compileDeleteRole(s *ir.DeleteRoleStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil delete role statement")
	}
	result := fmt.Sprintf(`DELETE ROLE "%s"`, s.RoleName)
	if s.Force {
		result += " FORCE"
	}
	return result + ";", nil
}

func compileGrantPermission(s *ir.GrantPermissionStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil grant permission statement")
	}
	return fmt.Sprintf(`GRANT "%s" TO USER "%s";`, s.Permission, s.Username), nil
}

func compileGrantRole(s *ir.GrantRoleStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil grant role statement")
	}
	return fmt.Sprintf(`GRANT "%s" TO USER "%s";`, s.RoleName, s.Username), nil
}

func compileRevokePermission(s *ir.RevokePermissionStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil revoke permission statement")
	}
	return fmt.Sprintf(`REVOKE "%s" FROM USER "%s";`, s.Permission, s.Username), nil
}

func compileRevokeRole(s *ir.RevokeRoleStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil revoke role statement")
	}
	return fmt.Sprintf(`REVOKE "%s" FROM USER "%s";`, s.RoleName, s.Username), nil
}

func compileUseDatabase(s *ir.UseDatabaseStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil use database statement")
	}
	return fmt.Sprintf(`USE DATABASE "%s";`, s.Database), nil
}

func compileCreateDatabase(s *ir.CreateDatabaseStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil create database statement")
	}
	return fmt.Sprintf(`CREATE DATABASE "%s";`, s.Database), nil
}

func compileDropDatabase(s *ir.DropDatabaseStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil drop database statement")
	}
	result := fmt.Sprintf(`DROP DATABASE "%s"`, s.Database)
	if s.Force {
		result += " FORCE"
	}
	return result + ";", nil
}

func compileRenameDatabase(s *ir.RenameDatabaseStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil rename database statement")
	}
	return fmt.Sprintf(`RENAME DATABASE "%s" TO "%s";`, s.OldName, s.NewName), nil
}

func compileAttachDatabase(s *ir.AttachDatabaseStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil attach database statement")
	}
	result := fmt.Sprintf(`ATTACH DATABASE "%s"`, s.Path)
	if s.Alias != "" {
		result += fmt.Sprintf(` AS "%s"`, s.Alias)
	}
	return result + ";", nil
}

func compileShowBundle(s *ir.ShowBundleStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil show bundle statement")
	}
	return fmt.Sprintf(`SHOW BUNDLE "%s";`, s.Bundle), nil
}

func compileShowViews(s *ir.ShowViewsStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil show views statement")
	}
	if s.Database != "" {
		return fmt.Sprintf(`SHOW VIEWS IN DATABASE "%s";`, s.Database), nil
	}
	return "SHOW VIEWS;", nil
}

func compileDescribeView(s *ir.DescribeViewStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil describe view statement")
	}
	return fmt.Sprintf(`DESCRIBE VIEW "%s";`, s.ViewName), nil
}

func compileShowMigrations(s *ir.ShowMigrationsStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil show migrations statement")
	}
	return fmt.Sprintf(`SHOW MIGRATIONS FOR "%s";`, s.Database), nil
}

func compileShowVersions(s *ir.ShowVersionsStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil show versions statement")
	}
	var b strings.Builder
	b.WriteString("SHOW VERSIONS")
	if s.DocumentID != "" {
		b.WriteString(fmt.Sprintf(` FOR "%s"`, s.DocumentID))
	}
	if s.Bundle != "" {
		b.WriteString(fmt.Sprintf(` IN BUNDLE "%s"`, s.Bundle))
	}
	b.WriteString(";")
	return b.String(), nil
}

func compileExplainSelect(s *ir.ExplainSelectStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil explain select statement")
	}
	inner := compileSelectInner(s.Query)
	return "EXPLAIN " + inner + ";", nil
}

func compileStartMigration(s *ir.StartMigrationStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil start migration statement")
	}
	if s.Description != "" {
		return fmt.Sprintf(`START MIGRATION WITH DESCRIPTION "%s";`, s.Description), nil
	}
	return "START MIGRATION;", nil
}

func compileApplyMigration(s *ir.ApplyMigrationStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil apply migration statement")
	}
	return fmt.Sprintf(`APPLY MIGRATION "%s";`, s.MigrationID), nil
}

func compileValidateMigration(s *ir.ValidateMigrationStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil validate migration statement")
	}
	return fmt.Sprintf(`VALIDATE MIGRATION "%s";`, s.MigrationID), nil
}

func compileValidateRollback(s *ir.ValidateRollbackStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil validate rollback statement")
	}
	return fmt.Sprintf(`VALIDATE ROLLBACK "%s";`, s.MigrationID), nil
}

func compileBackupDatabase(s *ir.BackupDatabaseStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil backup database statement")
	}
	result := fmt.Sprintf(`BACKUP DATABASE "%s"`, s.Database)
	if s.Path != "" {
		result += fmt.Sprintf(` TO "%s"`, s.Path)
	}
	return result + ";", nil
}

func compileRestoreDatabase(s *ir.RestoreDatabaseStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil restore database statement")
	}
	result := fmt.Sprintf(`RESTORE DATABASE FROM "%s"`, s.Path)
	if s.Database != "" {
		result += fmt.Sprintf(` AS "%s"`, s.Database)
	}
	return result + ";", nil
}

func compileInvalidateSession(s *ir.InvalidateSessionStmt) (string, error) {
	if s == nil {
		return "", fmt.Errorf("nil invalidate session statement")
	}
	return fmt.Sprintf(`INVALIDATE SESSION "%s";`, s.SessionID), nil
}

File: syndrql-ir-compiler/src/compiler.ts
import {
  Statement, IREnvelope, SelectStmt
} from './generated/types';
import { compileExpr, compileSelectInner } from './expression-compiler';

export function compile(stmt: Statement): string {
  switch (stmt.statementType) {
    // DML
    case 'select': return compileSelect(stmt.select!);
    case 'insert': return compileInsert(stmt);
    case 'update': return compileUpdate(stmt);
    case 'delete': return compileDelete(stmt);

    // DDL - Bundle
    case 'createBundle': return compileCreateBundle(stmt);
    case 'dropBundle': return compileDropBundle(stmt);
    case 'updateBundle': return compileUpdateBundle(stmt);

    // DDL - Index
    case 'createIndex': return compileCreateIndex(stmt);

    // DDL - View
    case 'createView': return compileCreateView(stmt);
    case 'dropView': return compileDropView(stmt);
    case 'refreshView': return `REFRESH MATERIALIZED VIEW "${stmt.refreshView!.viewName}";`;

    // Transaction
    case 'beginTransaction': return 'BEGIN TRANSACTION;';
    case 'commit': return 'COMMIT;';
    case 'rollback': return 'ROLLBACK;';
    case 'savepoint': return `SAVEPOINT "${stmt.savepoint!.name}";`;
    case 'rollbackToSavepoint': return `ROLLBACK TO SAVEPOINT "${stmt.rollbackToSavepoint!.name}";`;

    // Cursor
    case 'declareCursor': return compileDeclareCursor(stmt);
    case 'fetchCursor': return compileFetchCursor(stmt);
    case 'closeCursor': return `CLOSE ${stmt.closeCursor!.cursorName};`;

    // Prepared
    case 'prepare': return compilePrepare(stmt);
    case 'execute': return `EXECUTE ${stmt.execute!.statementName};`;
    case 'deallocate': return `DEALLOCATE ${stmt.deallocate!.statementName};`;

    // RBAC
    case 'createUser': return `CREATE USER "${stmt.createUser!.username}" WITH PASSWORD "${stmt.createUser!.password}";`;
    case 'updateUser': {
      let s = `UPDATE USER "${stmt.updateUser!.username}" SET PASSWORD = "${stmt.updateUser!.newPassword}"`;
      if (stmt.updateUser!.force) s += ' FORCE';
      return s + ';';
    }
    case 'deleteUser': {
      let s = `DELETE USER "${stmt.deleteUser!.username}"`;
      if (stmt.deleteUser!.force) s += ' FORCE';
      return s + ';';
    }
    case 'createRole': {
      let s = `CREATE ROLE "${stmt.createRole!.roleName}"`;
      if (stmt.createRole!.description) s += ` WITH DESCRIPTION "${stmt.createRole!.description}"`;
      return s + ';';
    }
    case 'updateRole': {
      let s = `UPDATE ROLE "${stmt.updateRole!.roleName}" SET DESCRIPTION = "${stmt.updateRole!.newDescription}"`;
      if (stmt.updateRole!.force) s += ' FORCE';
      return s + ';';
    }
    case 'deleteRole': {
      let s = `DELETE ROLE "${stmt.deleteRole!.roleName}"`;
      if (stmt.deleteRole!.force) s += ' FORCE';
      return s + ';';
    }
    case 'grantPermission': return `GRANT "${stmt.grantPermission!.permission}" TO USER "${stmt.grantPermission!.username}";`;
    case 'grantRole': return `GRANT "${stmt.grantRole!.roleName}" TO USER "${stmt.grantRole!.username}";`;
    case 'revokePermission': return `REVOKE "${stmt.revokePermission!.permission}" FROM USER "${stmt.revokePermission!.username}";`;
    case 'revokeRole': return `REVOKE "${stmt.revokeRole!.roleName}" FROM USER "${stmt.revokeRole!.username}";`;

    // Database
    case 'useDatabase': return `USE DATABASE "${stmt.useDatabase!.database}";`;
    case 'createDatabase': return `CREATE DATABASE "${stmt.createDatabase!.database}";`;
    case 'dropDatabase': {
      let s = `DROP DATABASE "${stmt.dropDatabase!.database}"`;
      if (stmt.dropDatabase!.force) s += ' FORCE';
      return s + ';';
    }
    case 'renameDatabase': return `RENAME DATABASE "${stmt.renameDatabase!.oldName}" TO "${stmt.renameDatabase!.newName}";`;
    case 'attachDatabase': {
      let s = `ATTACH DATABASE "${stmt.attachDatabase!.path}"`;
      if (stmt.attachDatabase!.alias) s += ` AS "${stmt.attachDatabase!.alias}"`;
      return s + ';';
    }

    // Utility
    case 'showDatabases': return 'SHOW DATABASES;';
    case 'showBundles': return 'SHOW BUNDLES;';
    case 'showBundle': return `SHOW BUNDLE "${stmt.showBundle!.bundle}";`;
    case 'showUsers': return 'SHOW USERS;';
    case 'showSessions': return 'SHOW SESSIONS;';
    case 'showViews': {
      if (stmt.showViews?.database) {
        return `SHOW VIEWS IN DATABASE "${stmt.showViews.database}";`;
      }
      return 'SHOW VIEWS;';
    }
    case 'describeView': return `DESCRIBE VIEW "${stmt.describeView!.viewName}";`;
    case 'showMigrations': return `SHOW MIGRATIONS FOR "${stmt.showMigrations!.database}";`;
    case 'showRateLimit': return 'SHOW RATE LIMIT;';
    case 'showVersions': {
      let s = 'SHOW VERSIONS';
      if (stmt.showVersions?.documentId) s += ` FOR "${stmt.showVersions.documentId}"`;
      if (stmt.showVersions?.bundle) s += ` IN BUNDLE "${stmt.showVersions.bundle}"`;
      return s + ';';
    }
    case 'showActiveSnapshots': return 'SHOW ACTIVE SNAPSHOTS;';
    case 'showConflictLog': return 'SHOW CONFLICT LOG;';
    case 'explainSelect': return `EXPLAIN ${compileSelectInner(stmt.explainSelect!.query!)};`;

    // Migration
    case 'startMigration': {
      if (stmt.startMigration?.description) {
        return `START MIGRATION WITH DESCRIPTION "${stmt.startMigration.description}";`;
      }
      return 'START MIGRATION;';
    }
    case 'applyMigration': return `APPLY MIGRATION "${stmt.applyMigration!.migrationId}";`;
    case 'validateMigration': return `VALIDATE MIGRATION "${stmt.validateMigration!.migrationId}";`;
    case 'validateRollback': return `VALIDATE ROLLBACK "${stmt.validateRollback!.migrationId}";`;

    // Ops
    case 'checkpoint': return 'CHECKPOINT;';
    case 'backupDatabase': {
      let s = `BACKUP DATABASE "${stmt.backupDatabase!.database}"`;
      if (stmt.backupDatabase!.path) s += ` TO "${stmt.backupDatabase!.path}"`;
      return s + ';';
    }
    case 'restoreDatabase': {
      let s = `RESTORE DATABASE FROM "${stmt.restoreDatabase!.path}"`;
      if (stmt.restoreDatabase!.database) s += ` AS "${stmt.restoreDatabase!.database}"`;
      return s + ';';
    }
    case 'invalidateSession': return `INVALIDATE SESSION "${stmt.invalidateSession!.sessionId}";`;

    default:
      throw new Error(`Unknown statement type: ${stmt.statementType}`);
  }
}

export function compileAll(env: IREnvelope): string[] {
  return env.statements.map(s => compile(s));
}

function compileSelect(s: SelectStmt): string {
  return compileSelectInner(s) + ';';
}

function compileValue(v: { type: string; raw: string } | undefined): string {
  if (!v) return 'NULL';
  switch (v.type) {
    case 'string': return `"${v.raw}"`;
    case 'int': return v.raw;
    case 'float': return v.raw;
    case 'bool': return v.raw;
    case 'null': return 'NULL';
    default: return 'NULL';
  }
}

function compileInsert(stmt: Statement): string {
  const s = stmt.insert!;
  const fieldStrs = s.fields.map(f => `{${f.name} = ${compileValue(f.value)}}`);
  return `ADD DOCUMENT TO BUNDLE "${s.bundle}" WITH (${fieldStrs.join(', ')});`;
}

function compileUpdate(stmt: Statement): string {
  const s = stmt.update!;
  const fieldStrs = s.fields.map(f => `${f.name} = ${compileValue(f.value)}`);
  let result = `UPDATE DOCUMENTS IN BUNDLE "${s.bundle}" (${fieldStrs.join(', ')})`;
  if (s.confirmed) result += ' CONFIRMED';
  if (s.where) result += ' WHERE ' + compileExpr(s.where);
  return result + ';';
}

function compileDelete(stmt: Statement): string {
  const s = stmt.delete!;
  let result = `DELETE DOCUMENTS FROM "${s.bundle}"`;
  if (s.confirmed) result += ' CONFIRMED';
  if (s.where) result += ' WHERE ' + compileExpr(s.where);
  return result + ';';
}

function compileCreateBundle(stmt: Statement): string {
  const s = stmt.createBundle!;
  const fieldStrs = s.fields.map(f => {
    let str = `{"${f.name}", "${f.type}", ${f.required}, ${f.unique}`;
    if (f.defaultValue) str += ', ' + compileValue(f.defaultValue);
    return str + '}';
  });
  return `CREATE BUNDLE "${s.bundle}" WITH FIELDS (${fieldStrs.join(', ')});`;
}

function compileDropBundle(stmt: Statement): string {
  let result = `DROP BUNDLE "${stmt.dropBundle!.bundle}"`;
  if (stmt.dropBundle!.force) result += ' FORCE';
  return result + ';';
}

function compileUpdateBundle(stmt: Statement): string {
  const s = stmt.updateBundle!;
  let result = `UPDATE BUNDLE "${s.bundle}"`;
  if (s.newBundleName) result += ` RENAME TO "${s.newBundleName}"`;
  if (s.changes) {
    for (const c of s.changes) {
      switch (c.action) {
        case 'ADD':
          if (c.fieldDef) {
            result += ` ADD FIELD "${c.field}" {"${c.fieldDef.name}", "${c.fieldDef.type}", ${c.fieldDef.required}, ${c.fieldDef.unique}}`;
          }
          break;
        case 'REMOVE':
          result += ` REMOVE FIELD "${c.field}"`;
          break;
        case 'MODIFY':
          if (c.fieldDef) {
            result += ` MODIFY FIELD "${c.field}" {"${c.fieldDef.name}", "${c.fieldDef.type}", ${c.fieldDef.required}, ${c.fieldDef.unique}}`;
          }
          break;
      }
    }
  }
  return result + ';';
}

function compileCreateIndex(stmt: Statement): string {
  const s = stmt.createIndex!;
  const fieldStrs = s.fields.map(f => `{"${f.name}", ${f.required}, ${f.unique}}`);
  return `CREATE ${s.indexType} "${s.indexName}" ON BUNDLE "${s.bundle}" WITH FIELDS (${fieldStrs.join(', ')});`;
}

function compileCreateView(stmt: Statement): string {
  const s = stmt.createView!;
  const prefix = s.isMaterialized ? 'CREATE MATERIALIZED VIEW' : 'CREATE VIEW';
  return `${prefix} "${s.viewName}" AS ${compileSelectInner(s.query!)};`;
}

function compileDropView(stmt: Statement): string {
  const s = stmt.dropView!;
  if (s.isMaterialized) return `DROP MATERIALIZED VIEW "${s.viewName}";`;
  return `DROP VIEW "${s.viewName}";`;
}

function compileDeclareCursor(stmt: Statement): string {
  const s = stmt.declareCursor!;
  return `DECLARE ${s.cursorName} CURSOR FOR ${compileSelectInner(s.query!)};`;
}

function compileFetchCursor(stmt: Statement): string {
  const s = stmt.fetchCursor!;
  if (s.fetchAll) return `FETCH ALL FROM ${s.cursorName};`;
  if (s.count === 1) return `FETCH NEXT FROM ${s.cursorName};`;
  return `FETCH ${s.count} FROM ${s.cursorName};`;
}

function compilePrepare(stmt: Statement): string {
  const s = stmt.prepare!;
  return `PREPARE ${s.statementName} AS ${compileSelectInner(s.query!)};`;
}
